Adding comments to Plan

Vielheim opened the pull request.

Commit:
Update fields in comments

Commit:
Update name to userId in Comment

Commit:
Add model for Comment

- Comment Model
- FirebaseAdaptedComment + Type
- CommentAdapter
- CommentController and FirebaseCommentController

Commit:
update FirebaseCommentController

Commit:
Add CommentViewModel

Currently only got one because add is together with update/delete/fetch. Planning to just have comments in one section on its own

Commit:
Add method to create CommentViewModel

The intention is that the comment is a child view under the Plan View, so the viewmodel will generate a new viewmodel for us.

Commit:
Create initial CommentsView

Commit:
Add MockCommentController

Commit:
Add MockUserController

Want to use to test rendering without Firebase

Commit:
Fix swiftlint

Commit:
Rename CommentViewModel to CommentsViewModel

Commit:
Add CommentView

Vielheim commented:
Add new collection: comments
- Comment model for Domain and Data representation
- New services for Comments

Add CommentsViewModel
- In charge of CRUD of comments
- Previously we separated the Add and Fetch/Edit/Delete into two VMs. But for comments because they are all embedded inside one View, I left the CRUD inside one VM. 


 Added Views:
- CommentsView: Parent view for each comment section.
- CommentListView: Renders a list of comments, sorted by date
- Individual CommentView. This view takes in the CommentsVM, a Comment and a User to render. I wanted to put the (Comment, User) inside a smaller CommentVM, but I am not able to get a "real-time" update whenever we edit or delete the comment. When I used the CommentsVM, I could decide when to re-render the entire Comments Section for a more real-time sort of update.
- EditCommentView - Create a `.sheet` like how we did for the other Models

Commit:
Add comments to other DetailViews

Commit:
Merge branch 'main' into plan-comments

# Conflicts:
#	TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift

Commit:
Add sorting for comments

Commit:
Add date formatting for Comment

Commit:
Add mock upvotes

Sort by latest

Commit:
Swap to FirebaseControllers

Commit:
Add check to see if comment is editable

Commit:
Update CommentView with edit and upvote

Commit:
Add method to upvote comment

Commit:
add todo - delete and edit

Commit:
Change to CommentsViewModel

Commit:
Add EditCommentView as a sheet

Didn't really want to do that, but didn't really have a choice. When we try to submit the changes via the CommentView, it will cause changes to the viewmodel which will re-render the view. Concurrently, we need to refresh the state of the view (to clear the edit fields). These two concurrent changes can cause unexpected behaviour to the view (based on the warnings). Hence, I used the sheet solution as per our other views, though it is a bit unsightly.

Commit:
Merge branch 'main' into plan-comments

# Conflicts:
#	TourMate/TourMate/Frontend/Plan/Services/CommentController.swift
#	TourMate/TourMate/Frontend/Plan/Services/FirebaseCommentController.swift
#	TourMate/TourMate/Frontend/Plan/Services/MockCommentController.swift
#	TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift
#	TourMate/TourMate/Frontend/User/Services/MockUserController.swift

Commit:
rename service

keanecjy commented on TourMate/TourMate/Backend/Persistence/Config/FirebaseConfig.swift:
```diff
@@ -12,6 +12,7 @@ final class FirebaseConfig {
     static let tripCollectionId = "trips"
     static let planCollectionId = "plans"
     static let planDetailsCollectionId = "plandetails"
```
can remove the plandetails above

keanecjy reviewed (commented):
1. Can bold the like button when the user has liked it so that the user can tell he has liked it.
2. The box gets bigger when the like button shows

Commit:
Merge branch 'main' into plan-comments-merge

# Conflicts:
#	TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift

Commit:
Update project file

Commit:
Update project file

Commit:
Add fix for transparent background

https://stackoverflow.com/questions/58696455/swiftui-ontapgesture-on-color-clear-background-behaves-differently-to-color-blue

Commit:
Change label for send

Commit:
Fix padding issue for Like button
- Have to add the same padding for the other views to achieve the same height

Fix logic bug in CommentsViewModel

Vielheim commented:
> * Can bold the like button when the user has liked it so that the user can tell he has liked it.
> * The box gets bigger when the like button shows

Fixed the box padding issue.
The bolding for text is not very obvious. But when you like the comment, the thumbsup icon is blue. If you haven't and someone else liked it, the icon is white.

keanecjy reviewed (approved):
LGTM
